LEGO1/mxdirectdraw.cpp

```diff
@@ -1,22 +1,1223 @@
+    if (result != DD_OK)
+    {
+      Error("SetClipper failed", result);
+      return FALSE;
+    }
+  }
+
+  return TRUE;
+}
+
+// OFFSET: LEGO1 0x1009D960
+BOOL MxDirectDraw::DDInit(BOOL fullscreen)
+{
+  HRESULT result;
+
+  if (fullscreen)
+  {
+    m_bIgnoreWM_SIZE = 1;
+    result = m_pDirectDraw->SetCooperativeLevel(m_hWndMain, DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN);
+    m_bIgnoreWM_SIZE = 0;
+  }
+  else
+  {
+    result = m_pDirectDraw->SetCooperativeLevel(m_hWndMain, DDSCL_NORMAL);
+  }
+
+  if (result != DD_OK)
+  {
+    Error("SetCooperativeLevel failed", result);
+    return FALSE;
+  }
+
+  m_bFullScreen = fullscreen;
+
+  return TRUE;
+}
+
+// OFFSET: LEGO1 0x1009DA80
+BOOL MxDirectDraw::DDSetMode(int width, int height, int bpp)
+{
+  HRESULT result;
+
+  if (m_bFullScreen)
+  {
+    LPDIRECTDRAW lpDD;
+
+    EnableResizing(m_hWndMain, FALSE);
+
+	if (!m_bIsOnPrimaryDevice)
+    {
+      lpDD = NULL;
+      result = DirectDrawCreate(0, &lpDD, 0);
+      if (result == DD_OK)
+      {
+        result = lpDD->SetCooperativeLevel(m_hWndMain, DDSCL_FULLSCREEN | DDSCL_EXCLUSIVE | DDSCL_ALLOWREBOOT);
+        if (result == DD_OK)
+        {
+          lpDD->SetDisplayMode(width, height, 8);
+        }
+      }
+    }
+
+    if (!IsSupportedMode(width, height, bpp))
+    {
+      width = m_pCurrentDeviceModesList->m_mode_ARRAY[0].width;
+      height = m_pCurrentDeviceModesList->m_mode_ARRAY[0].height;
+      bpp = m_pCurrentDeviceModesList->m_mode_ARRAY[0].bitsPerPixel;
+    }
+
+    m_bIgnoreWM_SIZE = TRUE;
+    result = m_pDirectDraw->SetDisplayMode(width, height, bpp);
+    m_bIgnoreWM_SIZE = FALSE;
+    if (result != DD_OK)
+    {
+      Error("SetDisplayMode failed", result);
+      return FALSE;
+    }
+  }
+  else
+  {
+    RECT rc;
+    DWORD dwStyle;
+
+    if (!m_bIsOnPrimaryDevice)
+    {
+      Error("Attempt made enter a windowed mode on a DirectDraw device that is not the primary display", E_FAIL);
+      return FALSE;
+    }
+
+    m_bIgnoreWM_SIZE = TRUE;
+    dwStyle = GetWindowLong(m_hWndMain, GWL_STYLE);
+    dwStyle &= ~(WS_POPUP | WS_CAPTION | WS_THICKFRAME | WS_OVERLAPPED);
+    dwStyle |= WS_CAPTION | WS_THICKFRAME | WS_OVERLAPPED;
+    SetWindowLong(m_hWndMain, GWL_STYLE, dwStyle);
+
+    SetRect(&rc, 0, 0, width - 1, height - 1);
+    AdjustWindowRectEx(
+      &rc,
+      GetWindowLong(m_hWndMain, GWL_STYLE),
+      GetMenu(m_hWndMain) != NULL,
+      GetWindowLong(m_hWndMain, GWL_EXSTYLE)
+    );
+    SetWindowPos(m_hWndMain, NULL, 0, 0, rc.right - rc.left + 1, rc.bottom - rc.top + 1, SWP_NOMOVE | SWP_NOZORDER | SWP_NOACTIVATE);
+    SetWindowPos(m_hWndMain, HWND_NOTOPMOST, 0, 0, 0, 0, SWP_NOSIZE | SWP_NOMOVE | SWP_NOACTIVATE);
+    m_bIgnoreWM_SIZE = FALSE;
+  }
+
+  m_currentMode.width = width;
+  m_currentMode.height = height;
+  m_currentMode.bitsPerPixel = bpp;
+
+  if (!DDCreateSurfaces())
+  {
+    return FALSE;
+  }
+
+  DDSURFACEDESC ddsd;
+
+  unk1();
```

Comment:
```suggestion
  FUN_1009E020();
```

Reply:
Perhaps a better name could be `MxDirectDraw::method_1009E020_ClearBackBuffersSlow`?

It's clearly a method (not a free-standing function), which is obvious by the caller initializing ecx and the callee using ecx.

If you look at the method it's not any fun at all - quite the opposite, e.g. `memset`-ing the backbuffer surface _line-by-line_.